base/MAPL_GridManager.F90

```diff
@@ -78,14 +79,83 @@ subroutine add_prototype(this, grid_type, prototype)
       
    end subroutine add_prototype
 
+   ! Is prototype_name present in the prototypes map keys?
+   logical function is_valid_prototype(this, prototype_name, unusable, rc)
+      class (GridManager), intent(inout) :: this
+      character(len=*), intent(in) :: prototype_name
+      class (KeywordEnforcer), optional, intent(in) :: unusable
+      integer, optional, intent(out) :: rc
+      integer :: status
+      logical, save :: initialized = .false.
```

Comment:
Possibly not threadsafe?   Might be ok though.

Reply:
In theory, it is not threadsafe. I _think_ in practice, it's OK.

`initialized` is a flag to prevent initializing the GridManager `prototypes` StringGridFactory map multiple times. The `make_clone` function was using for this purpose in existing code. In the new function `is_valid_prototype` and subroutine `initialize_prototypes`, I use it, as well. Originally, I added a `initialized` attribute to the GridManager for all three routines to set and get, but that definitely was not threadsafe. With the saved local variable, it's possible the value could vary by thread, which means the initialization could run extra times, but the state of the flag only changes from `.false.` to `.true.`.

The map itself would have greater thread safety issues, but since the initialization sets the entries in the map, multiple initializations should add the same entry to the map multiple times, overwriting each time. The initialization of the map is from constants. I don't know if the map ever changes after the initialization, but it appears to be a constant singleton-ish type for lookups. To be safe, we should probably limit a single thread to the initialization. We can discuss further.